File: ts/src/scheme.ts
export class URI {
  constructor(
    readonly scheme: string,
    readonly namespace: string,
    readonly id: string,
    readonly query = "",
    readonly fragment = "",
    readonly original = "",
    readonly action = ""
  ) {}

  canonical(): string {
    const query = this.query ? `?${this.query}` : "";
    const fragment = this.fragment ? `#${this.fragment}` : "";
    return `${this.scheme}://${this.namespace}/${this.id}${query}${fragment}`;
  }

  vanity(): string {
    return this.original || this.canonical();
  }

  toString(): string {
    return this.canonical();
  }
}

export type VanityAlias = {
  from: string;
  to: string;
  prefix?: boolean;
  preserveSuffix?: boolean;
};

export type ActionRoute = {
  command: string;
  args?: string[];
};

export type ResolvedAction = {
  action: string;
  command: string;
  args: string[];
};

export type Policy = {
  defaultNamespaceSegments?: number;
  schemeNamespaceSegments?: Record<string, number>;
  vanityAliases?: VanityAlias[];
  actionRoutes?: Record<string, ActionRoute>;
};

export type ParseOptions = {
  strict?: boolean;
  jsonAmbiguity?: boolean;
};

export type VanityCandidate = {
  from: string;
  to: string;
  distance: number;
};

export class AmbiguousVanityError extends Error {
  readonly input: string;
  readonly candidates: VanityCandidate[];

  constructor(input: string, candidates: VanityCandidate[], asJSON = false) {
    super(asJSON ? JSON.stringify({ input, candidates }) : `cite: ambiguous vanity alias ${JSON.stringify(input)}: ${candidates.map((c) => c.from).join(", ")}`);
    this.name = "AmbiguousVanityError";
    this.input = input;
    this.candidates = candidates;
  }

  toJSON(): { input: string; candidates: VanityCandidate[] } {
    return { input: this.input, candidates: this.candidates };
  }
}

export const defaultPolicy: Policy = {
  defaultNamespaceSegments: 1,
  schemeNamespaceSegments: {
    task: 2,
    doc: 2,
    repo: 1,
    tlc: 2,
    "task-dev": 2,
    "task-stress": 2,
  },
};

export const DefaultPolicy = defaultPolicy;

export function parse(input: string, policy: Policy = defaultPolicy, options: ParseOptions = {}): URI {
  if (input === "") throw new Error("cite: empty input");

  const resolved = resolveVanity(input, policy, options);
  const parsed = parseURL(resolved.parseInput);
  const namespaceSegmentCount = namespaceSegments(policy, parsed.scheme);
  if (namespaceSegmentCount <= 0) throw new Error("cite: namespace segment count must be positive");
  if (parsed.segments.length <= namespaceSegmentCount) throw new Error("cite: id is required");

  const namespace = parsed.segments.slice(0, namespaceSegmentCount).join("/");
  const id = parsed.segments.slice(namespaceSegmentCount).join("/");
  if (namespace === "") throw new Error("cite: namespace is required");
  if (id === "") throw new Error("cite: id is required");

  return new URI(parsed.scheme, namespace, id, parsed.query, parsed.fragment, resolved.vanity, actionFromQuery(parsed.query));
}

export function resolveAction(uri: URI, policy: Policy = defaultPolicy): ResolvedAction {
  if (!uri) throw new Error("cite: nil URI");
  if (!uri.action) throw new Error("cite: action is required");
  const route = policy.actionRoutes?.[uri.action];
  if (!route) throw new Error(`cite: unknown action ${JSON.stringify(uri.action)}`);
  if (!route.command) throw new Error("cite: action route command is required");

  return {
    action: uri.action,
    command: expandActionTemplate(route.command, uri),
    args: (route.args ?? []).map((arg) => expandActionTemplate(arg, uri)),
  };
}

export function vanityCandidates(input: string, policy: Policy = defaultPolicy): VanityCandidate[] {
  return (policy.vanityAliases ?? [])
    .map((alias) => ({ from: alias.from, to: alias.to, distance: levenshtein(input, alias.from) }))
    .filter((candidate) => withinFuzzyThreshold(input, candidate.from, candidate.distance))
    .sort((a, b) => (a.distance === b.distance ? a.from.localeCompare(b.from) : a.distance - b.distance));
}

function parseURL(input: string): { scheme: string; segments: string[]; query: string; fragment: string } {
  const schemeSep = input.indexOf("://");
  if (schemeSep <= 0) throw new Error("cite: scheme is required");

  const scheme = input.slice(0, schemeSep);
  let rest = input.slice(schemeSep + 3);
  let fragment = "";
  const hash = rest.indexOf("#");
  if (hash >= 0) {
    fragment = rest.slice(hash + 1);
    rest = rest.slice(0, hash);
  }

  let query = "";
  const queryIdx = rest.indexOf("?");
  if (queryIdx >= 0) {
    query = rest.slice(queryIdx + 1);
    rest = rest.slice(0, queryIdx);
  }

  const rawSegments = rest.split("/");
  const host = rawSegments.shift() ?? "";
  if (host === "") throw new Error("cite: namespace is required");

  const segments = [host];
  for (const segment of rawSegments) {
    if (segment === "") continue;
    try {
      segments.push(decodeURIComponent(segment));
    } catch (cause) {
      throw new Error(`cite: invalid path segment: ${(cause as Error).message}`);
    }
  }

  return { scheme, segments, query, fragment };
}

function actionFromQuery(rawQuery: string): string {
  let values: URLSearchParams;
  try {
    values = new URLSearchParams(rawQuery);
  } catch (cause) {
    throw new Error(`cite: invalid query: ${(cause as Error).message}`);
  }

  const candidates: string[] = [];
  const action = values.get("action") ?? "";
  const name = values.get("name") ?? "";
  if (action !== "" && name === "") candidates.push(action);

  const cmd = values.get("cmd") ?? "";
  const verb = values.get("verb") ?? "";
  if (cmd !== "" || verb !== "") {
    if (cmd === "" || verb === "") throw new Error("cite: cmd and verb must be provided together");
    candidates.push(`${cmd}.${verb}`);
  }

  if (name !== "") {
    if (action === "") throw new Error("cite: name and action must be provided together");
    candidates.push(`${name}.${action}`);
  }

  for (const candidate of candidates.slice(1)) {
    if (candidate !== candidates[0]) throw new Error("cite: conflicting action query parameters");
  }
  return candidates[0] ?? "";
}

function namespaceSegments(policy: Policy, scheme: string): number {
  const specific = policy.schemeNamespaceSegments?.[scheme];
  if (specific !== undefined) return specific;
  return policy.defaultNamespaceSegments && policy.defaultNamespaceSegments !== 0 ? policy.defaultNamespaceSegments : 1;
}

function resolveVanity(input: string, policy: Policy, options: ParseOptions): { parseInput: string; vanity: string } {
  let best: VanityAlias | undefined;
  let bestLen = -1;
  for (const candidate of policy.vanityAliases ?? []) {
    if (!candidate.from || !candidate.to) throw new Error("cite: vanity alias from and to are required");
    const matched = input === candidate.from || (candidate.prefix === true && input.startsWith(`${candidate.from}/`));
    if (matched && candidate.from.length > bestLen) {
      best = candidate;
      bestLen = candidate.from.length;
    }
  }

  if (!best) {
    if (!options.strict) {
      const fuzzy = closestVanity(input, policy, options);
      if (fuzzy) return fuzzy;
    }
    return { parseInput: input, vanity: "" };
  }

  let target = best.to;
  if (best.prefix && best.preserveSuffix && input.length > best.from.length) {
    target = best.to.replace(/\/+$/, "") + input.slice(best.from.length);
  }
  return { parseInput: target, vanity: input };
}

function closestVanity(input: string, policy: Policy, options: ParseOptions): { parseInput: string; vanity: string } | undefined {
  const candidates = vanityCandidates(input, policy);
  if (candidates.length === 0) return undefined;

  const bestDistance = candidates[0].distance;
  const best = candidates.filter((candidate) => candidate.distance === bestDistance);
  if (best.length > 1) throw new AmbiguousVanityError(input, best, options.jsonAmbiguity);
  return { parseInput: best[0].to, vanity: input };
}

function withinFuzzyThreshold(input: string, candidate: string, distance: number): boolean {
  const longest = Math.max([...input].length, [...candidate].length);
  const threshold = Math.min(8, Math.max(2, Math.floor(longest / 5)));
  return distance <= threshold;
}

function levenshtein(a: string, b: string): number {
  const ar = [...a];
  const br = [...b];
  if (ar.length === 0) return br.length;
  if (br.length === 0) return ar.length;

  let prev = Array.from({ length: br.length + 1 }, (_, i) => i);
  let curr = new Array<number>(br.length + 1).fill(0);
  for (let i = 0; i < ar.length; i++) {
    curr[0] = i + 1;
    for (let j = 0; j < br.length; j++) {
      const cost = ar[i] === br[j] ? 0 : 1;
      curr[j + 1] = Math.min(curr[j] + 1, prev[j + 1] + 1, prev[j] + cost);
    }
    [prev, curr] = [curr, prev];
  }
  return prev[br.length];
}

function expandActionTemplate(value: string, uri: URI): string {
  return value
    .replaceAll("{scheme}", uri.scheme)
    .replaceAll("{namespace}", uri.namespace)
    .replaceAll("{id}", uri.id)
    .replaceAll("{query}", uri.query)
    .replaceAll("{fragment}", uri.fragment);
}

File: rs/src/scheme.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::error::Error;
use std::fmt;
use url::Url;

#[derive(Debug, Clone, PartialEq, Eq, Default, Serialize, Deserialize)]
pub struct Uri {
    pub scheme: String,
    pub namespace: String,
    pub id: String,
    pub query: String,
    pub fragment: String,
    pub original: String,
    pub action: String,
}

impl Uri {
    pub fn canonical(&self) -> String {
        let mut out = format!("{}://{}/{}", self.scheme, self.namespace, self.id);
        if !self.query.is_empty() {
            out.push('?');
            out.push_str(&self.query);
        }
        if !self.fragment.is_empty() {
            out.push('#');
            out.push_str(&self.fragment);
        }
        out
    }

    pub fn vanity(&self) -> String {
        if self.original.is_empty() {
            self.canonical()
        } else {
            self.original.clone()
        }
    }
}

impl fmt::Display for Uri {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.canonical())
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Policy {
    pub default_namespace_segments: usize,
    pub scheme_namespace_segments: HashMap<String, usize>,
    pub vanity_aliases: Vec<VanityAlias>,
    pub action_routes: HashMap<String, ActionRoute>,
}

impl Default for Policy {
    fn default() -> Self {
        Self {
            default_namespace_segments: 1,
            scheme_namespace_segments: HashMap::from([
                ("task".to_string(), 2),
                ("doc".to_string(), 2),
                ("repo".to_string(), 1),
                ("tlc".to_string(), 2),
                ("task-dev".to_string(), 2),
                ("task-stress".to_string(), 2),
            ]),
            vanity_aliases: Vec::new(),
            action_routes: HashMap::new(),
        }
    }
}

impl Policy {
    pub fn namespace_segments(&self, scheme: &str) -> usize {
        self.scheme_namespace_segments
            .get(scheme)
            .copied()
            .or_else(|| {
                (self.default_namespace_segments != 0).then_some(self.default_namespace_segments)
            })
            .unwrap_or(1)
    }

    pub fn resolve_action(&self, uri: &Uri) -> Result<ResolvedAction, UriError> {
        if uri.action.is_empty() {
            return Err(UriError::ActionRequired);
        }
        let route = self
            .action_routes
            .get(&uri.action)
            .ok_or_else(|| UriError::UnknownAction(uri.action.clone()))?;
        if route.command.is_empty() {
            return Err(UriError::ActionRouteCommandRequired);
        }
        Ok(ResolvedAction {
            action: uri.action.clone(),
            command: expand_action_template(&route.command, uri),
            args: route
                .args
                .iter()
                .map(|arg| expand_action_template(arg, uri))
                .collect(),
        })
    }

    pub fn vanity_candidates(&self, input: &str) -> Vec<VanityCandidate> {
        let mut candidates: Vec<_> = self
            .vanity_aliases
            .iter()
            .filter_map(|alias| {
                let distance = levenshtein(input, &alias.from);
                within_fuzzy_threshold(input, &alias.from, distance).then(|| VanityCandidate {
                    from: alias.from.clone(),
                    to: alias.to.clone(),
                    distance,
                })
            })
            .collect();
        candidates.sort_by(|a, b| {
            a.distance
                .cmp(&b.distance)
                .then_with(|| a.from.cmp(&b.from))
        });
        candidates
    }

    fn resolve_vanity(
        &self,
        input: &str,
        options: ParseOptions,
    ) -> Result<(String, String), UriError> {
        let mut best: Option<&VanityAlias> = None;
        for alias in &self.vanity_aliases {
            if alias.from.is_empty() || alias.to.is_empty() {
                return Err(UriError::VanityAliasRequired);
            }
            if alias.to.contains("://") && Url::parse(&alias.to).is_err() {
                return Err(UriError::InvalidVanityTarget(alias.to.clone()));
            }
            let matched = input == alias.from
                || (alias.prefix && input.starts_with(&(alias.from.clone() + "/")));
            if matched && best.is_none_or(|current| alias.from.len() > current.from.len()) {
                best = Some(alias);
            }
        }

        if let Some(alias) = best {
            let mut target = alias.to.clone();
            if alias.prefix && alias.preserve_suffix && input.len() > alias.from.len() {
                target = format!(
                    "{}{}",
                    target.trim_end_matches('/'),
                    &input[alias.from.len()..]
                );
            }
            return Ok((target, input.to_string()));
        }

        if !options.strict {
            let candidates = self.vanity_candidates(input);
            if let Some(first) = candidates.first() {
                let best_distance = first.distance;
                let target = first.to.clone();
                let best: Vec<_> = candidates
                    .into_iter()
                    .take_while(|candidate| candidate.distance == best_distance)
                    .collect();
                if best.len() > 1 {
                    return Err(UriError::AmbiguousVanity(AmbiguousVanityError {
                        input: input.to_string(),
                        candidates: best,
                        as_json: options.json_ambiguity,
                    }));
                }
                return Ok((target, input.to_string()));
            }
        }

        Ok((input.to_string(), String::new()))
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct ParseOptions {
    pub strict: bool,
    pub json_ambiguity: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct VanityAlias {
    pub from: String,
    pub to: String,
    #[serde(default)]
    pub prefix: bool,
    #[serde(default, rename = "preserveSuffix")]
    pub preserve_suffix: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActionRoute {
    pub command: String,
    #[serde(default)]
    pub args: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ResolvedAction {
    pub action: String,
    pub command: String,
    pub args: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct VanityCandidate {
    pub from: String,
    pub to: String,
    pub distance: usize,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct AmbiguousVanityError {
    pub input: String,
    pub candidates: Vec<VanityCandidate>,
    pub as_json: bool,
}

impl fmt::Display for AmbiguousVanityError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if self.as_json {
            let candidates = self
                .candidates
                .iter()
                .map(|candidate| {
                    format!(
                        r#"{{"from":"{}","to":"{}","distance":{}}}"#,
                        json_escape(&candidate.from),
                        json_escape(&candidate.to),
                        candidate.distance
                    )
                })
                .collect::<Vec<_>>()
                .join(",");
            return write!(
                f,
                r#"{{"input":"{}","candidates":[{}]}}"#,
                json_escape(&self.input),
                candidates
            );
        }
        let names = self
            .candidates
            .iter()
            .map(|candidate| candidate.from.as_str())
            .collect::<Vec<_>>()
            .join(", ");
        write!(f, "cite: ambiguous vanity alias {:?}: {}", self.input, names)
    }
}

impl Error for AmbiguousVanityError {}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum UriError {
    EmptyInput,
    MissingScheme,
    EmptyNamespace,
    MissingId,
    InvalidInput(String),
    InvalidQuery(String),
    InvalidPathSegment(String),
    InvalidNamespaceSegmentCount,
    VanityAliasRequired,
    InvalidVanityTarget(String),
    AmbiguousVanity(AmbiguousVanityError),
    ConflictingActionQueryParameters,
    CmdAndVerbRequired,
    NameAndActionRequired,
    ActionRequired,
    UnknownAction(String),
    ActionRouteCommandRequired,
}

impl fmt::Display for UriError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::EmptyInput => f.write_str("cite: empty input"),
            Self::MissingScheme => f.write_str("cite: scheme is required"),
            Self::EmptyNamespace => f.write_str("cite: namespace is required"),
            Self::MissingId => f.write_str("cite: id is required"),
            Self::InvalidInput(err) => write!(f, "cite: invalid input: {err}"),
            Self::InvalidQuery(err) => write!(f, "cite: invalid query: {err}"),
            Self::InvalidPathSegment(err) => write!(f, "cite: invalid path segment: {err}"),
            Self::InvalidNamespaceSegmentCount => {
                f.write_str("cite: namespace segment count must be positive")
            }
            Self::VanityAliasRequired => f.write_str("cite: vanity alias from and to are required"),
            Self::InvalidVanityTarget(target) => write!(f, "cite: invalid vanity target: {target}"),
            Self::AmbiguousVanity(err) => err.fmt(f),
            Self::ConflictingActionQueryParameters => {
                f.write_str("cite: conflicting action query parameters")
            }
            Self::CmdAndVerbRequired => f.write_str("cite: cmd and verb must be provided together"),
            Self::NameAndActionRequired => {
                f.write_str("cite: name and action must be provided together")
            }
            Self::ActionRequired => f.write_str("cite: action is required"),
            Self::UnknownAction(action) => write!(f, "cite: unknown action {action:?}"),
            Self::ActionRouteCommandRequired => {
                f.write_str("cite: action route command is required")
            }
        }
    }
}

impl Error for UriError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        match self {
            Self::AmbiguousVanity(err) => Some(err),
            _ => None,
        }
    }
}

pub fn parse(input: &str) -> Result<Uri, UriError> {
    parse_with_policy_options(input, &Policy::default(), ParseOptions::default())
}

pub fn parse_with_policy(input: &str, policy: &Policy) -> Result<Uri, UriError> {
    parse_with_policy_options(input, policy, ParseOptions::default())
}

pub fn parse_with_policy_options(
    input: &str,
    policy: &Policy,
    options: ParseOptions,
) -> Result<Uri, UriError> {
    if input.is_empty() {
        return Err(UriError::EmptyInput);
    }

    let (parse_input, original) = policy.resolve_vanity(input, options)?;
    let parsed = Url::parse(&parse_input).map_err(|err| {
        if parse_input.contains("://") {
            UriError::InvalidInput(err.to_string())
        } else {
            UriError::MissingScheme
        }
    })?;

    if parsed.scheme().is_empty() {
        return Err(UriError::MissingScheme);
    }
    let host = parsed.host_str().ok_or(UriError::EmptyNamespace)?;
    if host.is_empty() {
        return Err(UriError::EmptyNamespace);
    }

    let mut segments = vec![host.to_string()];
    if let Some(path_segments) = parsed.path_segments() {
        for segment in path_segments.filter(|segment| !segment.is_empty()) {
            segments.push(segment.to_string());
        }
    }

    let namespace_segments = policy.namespace_segments(parsed.scheme());
    if namespace_segments == 0 {
        return Err(UriError::InvalidNamespaceSegmentCount);
    }
    if segments.len() <= namespace_segments {
        return Err(UriError::MissingId);
    }

    let namespace = segments[..namespace_segments].join("/");
    let id = segments[namespace_segments..].join("/");
    if namespace.is_empty() {
        return Err(UriError::EmptyNamespace);
    }
    if id.is_empty() {
        return Err(UriError::MissingId);
    }

    let query = parsed.query().unwrap_or_default().to_string();
    let action = action_from_query(&query)?;

    Ok(Uri {
        scheme: parsed.scheme().to_string(),
        namespace,
        id,
        query,
        fragment: parsed.fragment().unwrap_or_default().to_string(),
        original,
        action,
    })
}

fn action_from_query(raw_query: &str) -> Result<String, UriError> {
    let values = query_values(raw_query)?;
    let action = values.get("action").cloned().unwrap_or_default();
    let name = values.get("name").cloned().unwrap_or_default();
    let cmd = values.get("cmd").cloned().unwrap_or_default();
    let verb = values.get("verb").cloned().unwrap_or_default();

    let mut candidates = Vec::new();
    if !action.is_empty() && name.is_empty() {
        candidates.push(action.clone());
    }
    if !cmd.is_empty() || !verb.is_empty() {
        if cmd.is_empty() || verb.is_empty() {
            return Err(UriError::CmdAndVerbRequired);
        }
        candidates.push(format!("{cmd}.{verb}"));
    }
    if !name.is_empty() {
        if action.is_empty() {
            return Err(UriError::NameAndActionRequired);
        }
        candidates.push(format!("{name}.{action}"));
    }

    if candidates.is_empty() {
        return Ok(String::new());
    }
    if candidates
        .iter()
        .any(|candidate| candidate != &candidates[0])
    {
        return Err(UriError::ConflictingActionQueryParameters);
    }
    Ok(candidates.remove(0))
}

fn query_values(raw_query: &str) -> Result<HashMap<String, String>, UriError> {
    let mut values = HashMap::new();
    if raw_query.is_empty() {
        return Ok(values);
    }
    for pair in raw_query.split('&') {
        let (key, value) = pair.split_once('=').unwrap_or((pair, ""));
        values.insert(key.to_string(), value.to_string());
    }
    Ok(values)
}

fn expand_action_template(value: &str, uri: &Uri) -> String {
    value
        .replace("{scheme}", &uri.scheme)
        .replace("{namespace}", &uri.namespace)
        .replace("{id}", &uri.id)
        .replace("{query}", &uri.query)
        .replace("{fragment}", &uri.fragment)
}

fn within_fuzzy_threshold(input: &str, candidate: &str, distance: usize) -> bool {
    let longest = input.len().max(candidate.len());
    let threshold = (longest / 5).clamp(2, 8);
    distance <= threshold
}

fn levenshtein(a: &str, b: &str) -> usize {
    let ar: Vec<char> = a.chars().collect();
    let br: Vec<char> = b.chars().collect();
    if ar.is_empty() {
        return br.len();
    }
    if br.is_empty() {
        return ar.len();
    }

    let mut prev: Vec<usize> = (0..=br.len()).collect();
    let mut curr = vec![0; br.len() + 1];
    for (i, ac) in ar.iter().enumerate() {
        curr[0] = i + 1;
        for (j, bc) in br.iter().enumerate() {
            let cost = usize::from(ac != bc);
            curr[j + 1] = (curr[j] + 1).min(prev[j + 1] + 1).min(prev[j] + cost);
        }
        std::mem::swap(&mut prev, &mut curr);
    }
    prev[br.len()]
}

fn json_escape(value: &str) -> String {
    value.replace('\\', "\\\\").replace('"', "\\\"")
}
